### PR TITLE
Mail Service Impl

### LINKED ISSUE #370: Add an eMail Notification Service
Implement a service that allows the BPE to send emails on certain events.

### PR DESCRIPTION
**Pull request by hhund**
Adds a service implementation to send e-mails from DSF processes:
- If not configured, generated e-mail will be logged
- SMTP via TLS (smtps) can be enabled
- Username/Password or client-certificate authentication to the SMTP server can be configured
- A certificate trust store for the SMTP server can be configured
- A S/MIME certificate can be configured to send signed e-mails
- The BPE will send a test mail on startup if enabled
- The BPE will send mail with Log-Message on ERROR including the last 4 INFO and WARN Messages if enabled
- SMTP via HTTP or SOCKS proxy is currently **not** implemented

closes #370

### CODE REVIEW

File: dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java [hunk 1]
```diff
@@ -0,0 +1,400 @@
...
+			properties.put("mail.transport.protocol", "smtps");
+			properties.put("mail.smtp.socketFactory.fallback", "false");
+
+			properties.put("mail.smtp.ssl.checkserveridentity", "true");
+			properties.put("mail.smtp.ssl.socketFactory",
+					createSslSocketFactory(trustStore, keyStore, keyStorePassword));
+		}
+		else
+		{
+			properties.put("mail.smtp.host", mailServerHostname);
```
Comment (wetret) [inline dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java:197]: property set twice (L185/L197), could be moved to the top where the properties are initialized

File: dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java [hunk 2]
```diff
@@ -0,0 +1,400 @@
...
+			properties.put("mail.smtp.port", mailServerPort);
```
Comment (wetret) [inline dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java:198]: property set twice (L186/L198), could be moved to the top where the properties are initialized

File: dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/spring/config/PropertiesConfig.java [hunk 3]
```diff
@@ -174,6 +174,70 @@
...
+	@Value("${org.highmed.dsf.bpe.mail.trust.certificates:#{null}}")
+	private String mailServerTrustStoreFile;
+
+	@Documentation(description = "PEM encoded file with client certificate used to authenticate against the SMTP server. Requires SMTP over TLS to be enabled via *ORG_HIGHMED_DSF_BPE_MAIL_USESMTPS*", recommendation = "Use docker secret file to configure", example = "/run/secrets/smtp_server_client_certificate.pem")
+	@Value("${org.highmed.dsf.bpe.mail.client.certificate:#{null}}")
+	private String mailServerClientCertificateFile;
+
+	@Documentation(description = "Prvate key corresponging to the SMTP server client certificate as PEM encoded file. Use ${env_variable}_PASSWORD* or *${env_variable}_PASSWORD_FILE* if private key is encrypted. Requires SMTP over TLS to be enabled via *ORG_HIGHMED_DSF_BPE_MAIL_USESMTPS*", recommendation = "Use docker secret file to configure", example = "/run/secrets/smtp_server_client_certificate_private_key.pem")
```
Comment (wetret) [inline dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/spring/config/PropertiesConfig.java:221]: `Prvte` typo

**Review by wetret — COMMENTED** [review]
Thank you for the pull request, I have just three small comments.

Additionally it would be nice, if the proxy feature could be enabled as well. I think I know from at least one organization needing the feature.

File: dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/spring/config/PropertiesConfig.java (hunk 3 above)
Comment (hhund) [reply dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/spring/config/PropertiesConfig.java]: fixed with c19166d

File: dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java (hunk 2 above)
Comment (hhund) [reply dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java]: fixed with bdb79ce

File: dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java (hunk 1 above)
Comment (hhund) [reply dsf-bpe/dsf-bpe-server/src/main/java/org/highmed/dsf/bpe/service/SmtpMailService.java]: fixed with bdb79ce

**Review by wetret — APPROVED** [review]
I guess we can wait with the proxy implementation and if it is actually needed, we can add id in the future.

**Pull request merged**